datarepo-clienttests/build.gradle

```diff
@@ -96,6 +96,55 @@ task(runTest, dependsOn: 'classes', type: JavaExec) {
     main = "runner.TestRunner"
     classpath = sourceSets.main.runtimeClasspath
 }
+
+def getNamespace(){
+    def pathToServer = "src/main/resources/servers/" + System.env.TEST_RUNNER_SERVER_SPECIFICATION_FILE
+    println pathToServer
+    def serverFile = file(pathToServer)
+    def serverDefinition = new groovy.json.JsonSlurper().parseText(serverFile.text)
+    return serverDefinition.namespace
+}
+
+def runSecretScript(script){
+    def outputStream = new StringBuffer();
+    def proc = Runtime.getRuntime().exec(script)
+    proc.waitForProcessOutput(outputStream, System.err)
+    println(outputStream.toString());
+    return proc.exitValue()
+}
+
+task lockEnv {
+    doFirst {
+        if (System.env.TEST_RUNNER_SERVER_SPECIFICATION_FILE) {
+            def namespace = getNamespace()
+            def secretExitValue = runSecretScript("./tools/namespaceLock.sh $namespace")
+            if (secretExitValue > 0){
+                throw new GradleException("Lock could not be aquired.\n" +
+                    "REASON: Namespace '$namespace' already in use.\n" +
+                    "OVERRIDE: Run './gradlew unlock' to clear lock\n" +
+                    "(Warning! Check w/ team & github actions to assert no one else using namespace)")
+            }
+        } else {
+            throw new GradleException("TEST_RUNNER_SERVER_SPECIFICATION_FILE must be defined in order to lock deployment.")
+        }
+    }
+}
+
+task unLockEnv {
+    doLast {
+        if (System.env.TEST_RUNNER_SERVER_SPECIFICATION_FILE) {
+            def namespace = getNamespace()
+            def secretExitValue = runSecretScript("./tools/deleteNamespaceLock.sh $namespace")
+            if (secretExitValue > 0){
+                throw new GradleException("Failed to clear locks from namespace $namespace")
+            }
+        } else {
+            throw new GradleException("TEST_RUNNER_SERVER_SPECIFICATION_FILE must be defined in order to lock deployment.")
+        }
+    }
+}
+
+
```

Comment:
Here's some interesting code to play with to get the right behavior.  There's basically a catch-all task that runs:
- Lock
- Run Test (and finalized By makes it ensure that it runs unlock whether or not test succeeds

```task unLockEnv {
    doLast {
        println 'unLockEnv'
    }
}
task runTest {
    doLast {
        println 'runTest'
//        throw new RuntimeException("FAILED!!!")
    }
}
task lockEnv {
    doLast {
        println 'lockEnv'
    }
}
task runTest.mustRunAfter(lockEnv)
task unLockEnv.mustRunAfter(runTest)
task lockAndRunTest {
    runTest.finalizedBy(unLockEnv)
    dependsOn(lockEnv, runTest)
}
```

So with this, if you run:
```../gradlew -q lockAndRunTest``` you get
```lockEnv
runTest
unLockEnv```

Note: you can still run the tasks independently

You can play with adding/removing the `Throw`

Reply:
Thank you for the helpful code!!!